### PR TITLE
[SPARK-26416] [SQL] Refactor `ColumnPruning` from `Optimizer.scala` to `ColumnPruning.scala`

### PR DESCRIPTION
**Pull request by dbtsai**
## What changes were proposed in this pull request?

As `Optimizer.scala` becomes bigger and bigger, it's hard to add new rules and maintain them. We are refactoring out `ColumnPruning` from `Optimizer.scala` to `ColumnPruning.scala` so it's easier to add new logics in `ColumnPruning`.

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by dbtsai** [comment]
cc @gatorsmile @cloud-fan @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ColumnPruning.scala [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+  /**
+   * The Project before Filter is not necessary but conflict with PushPredicatesThroughProject,
+   * so remove it. Since the Projects have been added top-down, we need to remove in bottom-up
+   * order, otherwise lower Projects can be missed.
+   */
+  private def removeProjectBeforeFilter(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case p1 @ Project(_, f @ Filter(_, p2 @ Project(_, child)))
+        if p2.outputSet.subsetOf(child.outputSet) =>
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ColumnPruning.scala]: I can see that this is a clean refactoring and this line is the additional style fix.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (pening Jenkins)

**Review by gatorsmile — CHANGES_REQUESTED** [review]
I would not suggest to move the code with such a complex change history.

**Comment by dongjoon-hyun** [comment]
@gatorsmile . Spark 3.0 is a good chance to make this more flexible in the future.

**Comment by gatorsmile** [comment]
It is hard for us to do the review in the future. Normally, we check the change history when doing the review. The change history is very important for us to do the review.

**Comment by gatorsmile** [comment]
Discussed it with @dongjoon-hyun offline. There are various bug fixes in the code history. The change history is very valuable for us [especially for new contributors] to understand the code. Previously, we did the split, but it makes us much harder to find out the original PRs who introduced the changes. 

For example, the following screenshot can clearly show all the PRs that changed this column pruning rule. The change history is very useful for reviewers and new learners. 

screen shot 2018-12-20 at 6 10 41 pm

**Comment by gatorsmile** [comment]
Maybe not all of you know this trick. You can get the history by selection in IntelliJ. See the screen shot. 

screen shot 2018-12-20 at 6 15 09 pm

**Comment by dongjoon-hyun** [comment]
Yep. I agree with @gatorsmile now.

**Comment by dbtsai** [comment]
All the above is great conversation. I'm closing this PR. Thanks.

**Pull request closed without merging**